### PR TITLE
spec: add spec for notation verify command

### PR DESCRIPTION
**Pull request by yizha1**

### CODE REVIEW

File: specs/commandline/verify.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# notation verify
+
+## Description
+
+Use `notation verify` command to verify signatures on an artifact. Signature verification succeeds if verification succeeds for at least one signature. The digest of the supplied artifact is returned upon successful verification.
```
Comment (gokarnm) [inline specs/commandline/verify.md:5]: Feel free to reword further.

```suggestion
Use `notation verify` command to verify signatures on an artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. The digest of the supplied artifact is returned upon successful verification. It is recommended that this digest reference be used to pull the artifact subsequently, as registry tags may be mutable, and a tag reference can point to a different artifact that what was verified.
```

File: specs/commandline/verify.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+Use `notation certificate` command to configure trust stores.
+
+### Configure Trust Policy
+
+Users who consume signed artifact from a registry use the trust policy to specify trusted identities which sign the artifacts, and level of signature verification to use. The trust policy is a JSON document. User needs to create a file named `trustpolicy.json` under `{NOTATION_CONFIG}`. See [Notation Directory Structure](https://github.com/notaryproject/notation/blob/main/specs/directory.md) for `{NOTATION_CONFIG}`.
+
+An example of `trustpolicy.json`:
+
+```text
```
Comment (gokarnm) [inline specs/commandline/verify.md:33]:
```suggestion
```jsonc
```

File: specs/commandline/verify.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+              "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder"
+            ]
+        }
+    ]
+}
+```
+
+In this example, only one policy is configured with the name `wabbit-networks-images`. With the value of property `registryScopes` set to `*`, this policy applies to all artifacts from any registry location. User can configure multiple trust policies for different scenarios. See [Trust Policy Schema and properties](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-policy) for details.
+
+### Verify signatures on a container image stored in a registry (Neither trust store nor trust policy is configured)
```
Comment (gokarnm) [inline specs/commandline/verify.md:55]:
```suggestion
### Verify signatures on a container image stored in a registry (Neither trust store nor trust policy is configured yet)
```

**Review by gokarnm — CHANGES_REQUESTED** [review]
Thanks @yizha1! Added some feedback.

File: specs/commandline/verify.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
+# notation verify
+
+## Description
+
+Use `notation verify` command to verify signatures on an artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. The digest of the supplied artifact is returned upon successful verification. It is recommended that this digest reference be used to pull the artifact subsequently, as registry tags may be mutable, and a tag reference can point to a different artifact that what was verified.
+
+## Outline
+
+```text
+Verify artifacts against signatures
```
Comment (priteshbandi) [inline specs/commandline/verify.md:10]:
```suggestion
Verifies signatures associated with the artifact.
```

File: specs/commandline/verify.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+Flags:
+  -h, --help                    help for verify
+  -p, --password string         Password for registry operations (default to $NOTATION_PASSWORD if not specified)
+      --plugin-config strings   {key}={value} pairs that are passed as is to a plugin, if the verification is associated with a verification plugin, refer plugin documentation to set appropriate values
+  -u, --username string         Username for registry operations (default to $NOTATION_USERNAME if not specified)
+```
+
+## Usage
+
+User needs to configure trust store and trust policy properly before using `notation verify` command.
```
Comment (priteshbandi) [inline specs/commandline/verify.md:24]:
```suggestion

Pre-requisite: User needs to configure trust store and trust policy properly before using `notation verify` command.
```

**Review by priteshbandi — APPROVED** [review]
LGTM with below suggestions.

File: specs/commandline/verify.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+Use `notation verify` command to verify signatures on an artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. The digest of the supplied artifact is returned upon successful verification. It is recommended that this digest reference be used to pull the artifact subsequently, as registry tags may be mutable, and a tag reference can point to a different artifact that what was verified.
+
+## Outline
+
+```text
+Verify signatures associated with the artifact.
+
+Usage:
+  notation verify <reference> [flags]
```
Comment (shizhMSFT) [inline specs/commandline/verify.md:13]:
```suggestion
  notation verify [flags] <reference>
```

File: specs/commandline/verify.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            "signatureVerification": {          // The level of verification - strict, permissive, audit, skip.
+              "level" : "strict" 
+            },
+            "trustStores": ["ca:wabbit-networks"], // The trust stores that contains the X.509 trusted roots.
+            "trustedIdentities": [              // Identities that are trusted to sign the artifact.
+              "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder"
+            ]
+        }
+    ]
+}
```
Comment (shizhMSFT) [inline specs/commandline/verify.md:53]: nit: formatting requested.
```suggestion
{
    "version": "1.0",
    "trustPolicies": [
        {
            // Policy for all artifacts, from any registry location.
            "name": "wabbit-networks-images", // Name of the policy.
            "registryScopes": [ "*" ],        // The registry artifacts to which the policy applies.
            "signatureVerification": {        // The level of verification - strict, permissive, audit, skip.
                "level": "strict"
            },
            "trustStores": [ "ca:wabbit-networks" ], // The trust stores that contains the X.509 trusted roots.
            "trustedIdentities": [                   // Identities that are trusted to sign the artifact.
                "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder"
            ]
        }
    ]
}
```

File: specs/commandline/verify.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
```
Comment (shizhMSFT) [inline specs/commandline/verify.md:85]: nit: formatting requested.
```suggestion
{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "wabbit-networks-images", // Name of the policy.
            "registryScopes": [ "registry.wabbit-networks.io/software/net-monitor" ], // The registry artifacts to which the policy applies.
            "signatureVerification": {        // The level of verification - strict, permissive, audit, skip.
                "level": "strict"
            },
            "trustStores": [ "ca:wabbit-networks" ], // The trust stores that contains the X.509 trusted roots.
            "trustedIdentities": [                   // Identities that are trusted to sign the artifact.
                "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Finance, CN=SecureBuilder"
            ]
        }
    ]
}
```

File: specs/commandline/verify.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+EOF
+
+# Verify signatures on a container image 
+notation verify registry.wabbit-networks.io/software/net-monitor:v1
+
```
Comment (shizhMSFT) [inline specs/commandline/verify.md:90]: nit: remove empty line

File: specs/commandline/verify.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+### Verify signatures on an OCI artifact stored in a registry (Trust store and trust policy are configured properly)
+
+```shell
+# Prerequisites: Signatures are stored in a registry referencing the signed OCI artifact
+
+# Verify signatures on an OCI artifact identified by the digest
+notation verify registry.wabbit-networks.io/software/net-monitor@sha256:abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234
+
```
Comment (shizhMSFT) [inline specs/commandline/verify.md:100]: nit: remove empty line

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by gokarnm — APPROVED** [review]
LGTM!

File: specs/commandline/verify.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    ]
+}
+EOF
+
+# Verify signatures on a container image 
+notation verify registry.wabbit-networks.io/software/net-monitor:v1
+```
+
+### Verify signatures on an OCI artifact stored in a registry (Trust store and trust policy are configured properly)
```
Comment (dtzar) [inline specs/commandline/verify.md:92]: This is confusing as to OCI artifact versus trust * configured versus not.
I'd consider removing the distinction of OCI artifact per section.  Place the tag AND digest verification options for image & OCI artifact in the same properly configured section.  

I'd then have a section on what the experience is like when trust store and policy is not configured.  I'd imagine you wouldn't get a verification, but show some error message(s).

File: specs/commandline/verify.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+```shell
+
+# Prerequisites: Signatures are stored in a registry referencing the signed container image
+
+# Configure trust store by adding a certificate file into trust store named "wabbit-network" of type "ca"
+notation certificate add --type ca --store wabbit-networks wabbit-networks.crt
+
+# Configure trust policy by creating a JSON document named "trustpolicy.json" under directory "{NOTATION_CONFIG}"
+# Example on Linux
+cat <<EOF > $HOME/.config/notation/trustpolicy.json
```
Comment (dtzar) [inline specs/commandline/verify.md:69]: This is going to vary based on the user's operating system.  IMO we should consider if we can get a "bare bones" cli implementation which simply writes a template similar to what's there to the user's proper directory (i.e. **notation policy create -n default**) and possibly opens it up or lists where it is at.  (i.e. **notation policy list** or **notation policy open** (optional **--name default**)

File: specs/commandline/verify.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+Use `notation verify` command to verify signatures on an artifact. Signature verification succeeds if verification succeeds for at least one of the signatures associated with the artifact. The digest of the supplied artifact is returned upon successful verification. It is recommended that this digest reference be used to pull the artifact subsequently, as registry tags may be mutable, and a tag reference can point to a different artifact that what was verified.
+
+## Outline
+
+```text
+Verify signatures associated with the artifact.
+
+Usage:
+  notation verify [flags] <reference>
```
Comment (dtzar) [inline specs/commandline/verify.md:13]: We should explain here and in the sign command that < reference > support can be a tag or a digest.  When with a tag, we do a default tag to digest translation and we only sign/verify the digest, not the tag itself.

**Review by dtzar — APPROVED** [review]
Open to address these before or after merge, so mark for approve.

File: specs/commandline/verify.md (hunk 13 above)
Comment (yizha1) [reply specs/commandline/verify.md:13]: I will create a new PR to address this

File: specs/commandline/verify.md (hunk 11 above)
Comment (yizha1) [reply specs/commandline/verify.md:92]: I will create a new PR to address this comment and the error handling parts.

File: specs/commandline/verify.md (hunk 12 above)
Comment (yizha1) [reply specs/commandline/verify.md:69]: I will create a new issue for a discussion regarding this request.

Comment (yizha1) [reply specs/commandline/verify.md:69]: https://github.com/notaryproject/notation/issues/398

**Pull request merged**